ordered-imports rule

danvk opened the pull request:
Fixes #771

See test & rule metadata for how this works. No options for now, but I could imagine many different ways to order & group imports.

thasner commented on test/rules/ordered-imports/test.ts.lint:
```diff
@@ -0,0 +1,21 @@
+
+import * as foo from 'foo';
+import * as bar from 'bar'; // failure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~            [ordered-sources]
```
consider adding a test for blank-lines resetting the `lastImportSource`

thasner commented:
👍

danvk commented:
I added a line to the test that would break if `lastImportSource` weren't reset.

danvk commented:
@adidahiya mind taking a look at this? Thanks!

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,125 @@
+}
+
+// Returns "./utils" given the tree for, e.g. "import * as utils from "./utils";"
+function findFromSource(node: ts.Node): string {
```
Instead of doing all this (which seems correct to me btw) you should be able to do something like `node.moduleSpecifier.getText()`. (Assuming `node` is a `ts.ImportDeclaration`.)

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,125 @@
+            continue;
+        }
+        const next = node.getChildAt(i + 1);
+        if (next.kind !== ts.SyntaxKind.StringLiteral) {
```
I appreciate this explicit handling of errors, but it turns out its normally better to silently continue if something malformed is found. Reason being that when people use TSLint in their editor, their code gets linted as they type and the resulting AST isn't always completely well-formed. We don't really want TSLint to start throwing errors here though, it's better for it to just move on.

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,125 @@
+    // This is the "{A, B, C}" of "import {A, B, C} from "./foo";".
+    // We need to make sure they're alphabetized.
+    public visitNamedImports(node: ts.NamedImports) {
+        const names = node.getChildAt(1);    // Three children: "{", "A, B, C" and "}"
```
`node` here will have a field called `elements`, which is an array of `ts.ImportSpecifier`

JKillian commented on test/rules/ordered-imports/test.ts.lint:
```diff
@@ -0,0 +1,22 @@
+import * as foo from 'foo';
+import * as bar from 'bar'; // failure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~            [ordered-sources]
+
```
Could we add a test case with this kind of import:
```
import someDefault from "module";
```

as well as the combined form:

```
import someDefault, {nameA, nameB: differentIdentifier} from "./wherever";
```

Just is a nice sanity check that the rule doesn't throw errors or struggle with all the different forms of imports

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,125 @@
+        }
+
+        const pair = findUnsortedPair(imports);
+        if (pair) {
```
Small stylistic note: we like to explicitly do `pair != null` since it avoids the unexpected quirks of JS type conversions.

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,125 @@
+        const prefixLength = node.getStart() - node.getFullStart();
+        const prefix = node.getFullText().slice(0, prefixLength);
+
+        if (prefix.indexOf("\n\n") >= 0) {
```
Will this work on files with `\r\n` linebreaks?

JKillian commented:
This PR looks very good, thanks @danvk! I left a few comments above for tweaks/improvements.

My first thoughts were that `import {SomeClass, someFunc} from "a"` should be enforced instead of `import {someFunc, SomeClass} from "a"`. Perhaps this is one of the cases though where it doesn't really matter which is the default, as long as users can choose which they want (eventually). There do seem to be many possible options for this rule, but perhaps we're better off starting simple as you've implemented it here.

danvk commented:
@JKillian Thanks for the educational review! I've taken all your suggestions and added an option to control the order of named declarations. Valid options are:

  * `"case-insensitive"`: Correct order is `{A, b, C}`. (This is the default.)
  * `"lowercase-first"`: Correct order is `{b, A, C}`.
  * `"lowercase-last"`: Correct order is `{A, C, b}`.

For named imports like:

```
import {longName as name} from 'foo';
import {name: longName} from 'foo';
```

I'm sorting by the full text, i.e. `longName as name` and `name: longName`.

Let me know if you have any more suggestions.

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,137 @@
+            type: "object",
+            properties: {
+                "named-imports-order": {
+                    oneOf: ["case-insensitive", "lowercase-first", "lowercase-last"],
```
I don't believe this is valid JSO Schema. I think you'll want this:

```
"named-imports-order": {
    type: "string",
    enum: ["case-insensitive", "lowercase-first", "lowercase-last"]
}
```

JKillian commented:
My apologies for my one comment, I don't believe `import { symbol: localName } from "./wherever";` is valid ES6 or valid TS! Using `as` as you mention above is correct, and I think the way you handle it currently is good.

(As as aside, I wonder why the syntax differs between importing and destructuring?)

Those options you provide seem good to me. It also keeps the door open for adding ordering of groups in the future.

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,137 @@
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static NAMED_IMPORTS_UNORDERED = "Named imports must be alphabetized.";
+    public static IMPORT_SOURCES_UNORDERED = "Import sources within a block must be alphabetized.";
```
Since we use the word "group" in the documentation above, lets also use it here. I'm afraid "block" will imply the `{  }` type of block

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,137 @@
+            - Import sources must be alphabetized within groups, i.e.:
+                    import * as a from "a";
+                    import * as b from "b";
+            - groups of imports are delineated by blank lines. You can use these to group imports
```
Let's capitalize "Groups" to be consistent with the above bullet points

JKillian commented on src/rules/orderedImportsRule.ts:
```diff
@@ -0,0 +1,137 @@
+                - The exact ordering can be controled by the named-imports-order option.
+                - "longName as name" imports are ordered by "longName".
+            - Import sources must be alphabetized within groups, i.e.:
+                    import * as a from "a";
```
Can we change this to:

```
import * as foo from "a";
import * as bar from "b";
```

so users can more easily distinguish what they should sort on?

JKillian commented:
Looking good! Just left a few comments on some minor docs changes

danvk commented:
Addressed all comments.

I'm not sure why the syntaxes for imports and destructuring are different. I certainly find the `{name as localName}` syntax to be much clearer. I never use `const {a: b} = o;` because I can never remember whether that means `const a = o.b;` or `const b = o.a;`.

JKillian commented:
Sweet, merged!

glen-84 commented:
I order my imports based on the full text, and not the "source". Two benefits are:

1. I can use something like "sort lines" in my editor (VS Code).
2. It's easier to look up a named import (for example) in this alphabetical order. The "source" is possibly less important.

I'm not 100% sure about this though. I may change my mind. :stuck_out_tongue:
